QubitServer/src/main/java/org/labrad/qubits/config/AdcDemodConfig.scala

```diff
@@ -0,0 +1,213 @@
+    require(Is.length == Qs.length, "Is and Qs must be of the same shape!")
+    require(inUse(channel), s"Interpret phases on channel $channel -- channel not turned on!")
+    //System.out.println("interpretPhases: channel " + channel + " crit phase: " + criticalPhase[channel]);
+    (Is zip Qs).map { case (i, q) =>
+      Math.atan2(q, i) > criticalPhase(channel)
+    }
+  }
+
+  def setFilterFunction(filterFunction: String, stretchLen: Int, stretchAt: Int): Unit = {
+    this.filterFunction = filterFunction
+    this.stretchLen = stretchLen
+    this.stretchAt = stretchAt
+  }
+
+  def setTrigMagnitude(channel: Int, ampSin: Int, ampCos: Int): Unit = {
+    require(channel <= MAX_CHANNELS, s"channel must be <= $MAX_CHANNELS")
+    require(ampSin > -1 && ampSin <= TRIG_AMP && ampCos > -1 && ampCos <= TRIG_AMP,
+        s"Trig Amplitudes must be 0-255 for channel '$name'")
+    this.inUse(channel) = true
+    this.ampSin(channel) = ampSin
+    this.ampCos(channel) = ampCos
+  }
+
+  /**
+   * sets the demodulation phase
+   * @param channel the channel index
+   * @param dPhi the number of addresses to step through per time step
+   * @param phi0 the initial offset
+   */
+  def setPhase(channel: Int, dPhi: Int, phi0: Int): Unit = {
+    require(channel <= MAX_CHANNELS, s"channel must be <= $MAX_CHANNELS")
+    //Preconditions.checkArgument(phi0 >= 0 && phi0 < (int)Math.pow(2, LOOKUP_ACCUMULATOR_BITS),
+    //"phi0 must be between 0 and 2^%s", LOOKUP_ACCUMULATOR_BITS);
+    inUse(channel) = true
+    this.dPhi(channel) = dPhi
+    this.phi0(channel) = phi0
+  }
+
+  /**
+   * sets the demodulation phase.
+   * @param channel the channel index
+   * @param frequency in Hz. it is converted in this function.
+   * @param phase of the offset IN RADIANS. it is converted to an address.
+   */
+  def setPhase(channel: Int, frequency: Double, phase: Double): Unit = {
+    require(phase >= -Math.PI && phase <= Math.PI, "Phase must be between -pi and pi")
+    val dPhi = Math.floor(frequency * Math.pow(2, LOOKUP_ACCUMULATOR_BITS) * DEMOD_TIME_STEP * Math.pow(10, -9.0)).toInt
+    val phi0 = (phase * Math.pow(2, LOOKUP_ACCUMULATOR_BITS) / (2 * Math.PI)).toInt
+    setPhase(channel, dPhi, phi0)
+  }
+
+  /**
+   * In the demod case, the following packets are added:
+   * ADC Run Mode
+   * Start Delay
+   * ADC Filter Func
+   * for each channel: ADC Demod Phase
+   *                   ADC Trig Magnitude
+   * @param runRequest The request to which we add the packets.
+   * @author pomalley
+   */
+  def addPackets(runRequest: Request): Unit = {
+    // check that the user has set everything that needs to be set
+    require(startDelay > -1, s"ADC Start Delay not set for channel '$name'")
+    require(stretchLen > -1 && stretchAt > -1, s"ADC Filter Func not set for channel '$name'")
+    require(inUse.exists(_ == true), s"No demod channels activated for channel '$name'")
```

Comment:
intellij says to use contains instead of exists. contains reads a little cleaner to me but mostly because i'm not that familiar with scala yet.